services:
  datalake:
    image: minio/minio:RELEASE.2025-02-03T21-03-04Z-cpuv1
    volumes:
      - minio_data:/data
    networks:
      - datalakehouse
    environment:
      MINIO_ROOT_USER: ${MINIO_ROOT_USER}
      MINIO_ROOT_PASSWORD: ${MINIO_ROOT_PASSWORD}
    command: minio server /data --console-address ":9001"
    deploy:
      restart_policy:
        condition: on-failure
    ports:
      - "9000:9000"
      - "9001:9001"

  hive-postgres:
    image: postgres:15
    environment:
      POSTGRES_USER: hive
      POSTGRES_PASSWORD: password
      POSTGRES_DB: metastore_db
    networks:
      - datalakehouse
    ports:
      - "5432:5432"
    healthcheck:
      test: ["CMD", "pg_isready", "-U", "metastore_db"]
      interval: 10s
      timeout: 5s
      retries: 5

  hive-server:
    image: hive-with-pgdriver
    ports:
      - "10000:10000"
      - "10002:10002"
    environment:
      SERVICE_NAME: hiveserver2
      IS_RESUME: "true"
      DB_SERVICE_OPTS: "-Dhive.metastore.uris=thrift://metastore:9083"
    volumes:
      - warehouse:/opt/hive/data/warehouse
    networks:
      - datalakehouse

  hive-metastore:
    image: hive-with-pgdriver
    ports:
      - "9083:9083"
    environment:
      SERVICE_NAME: metastore
      DB_DRIVER: postgres
      DB_SERVICE_OPTS: "-Djavax.jdo.option.ConnectionDriverName=org.postgresql.Driver \
                        -Djavax.jdo.option.ConnectionURL=jdbc:postgresql://hive-postgres:5432/metastore_db \
                        -Djavax.jdo.option.ConnectionUserName=hive \
                        -Djavax.jdo.option.ConnectionPassword=password"
      HIVE_CUSTOM_CONF_DIR: "./hive_custom_conf"
    volumes:
      - warehouse:/opt/hive/data/warehouse
      - ./hive_custom_conf/:/opt/hive/conf
    networks:
      - datalakehouse
    depends_on:
      hive-postgres:
        condition: service_healthy

networks:
  datalakehouse:
    external: true
    driver: bridge

volumes:
  minio_data:
  warehouse:
